### PR TITLE
Introduce a super inefficient per class dataloader

### PR DESCRIPTION
**Pull request by francescodeaglio**
To support some downsampling strategies, we need to be able to get samples label by label. In this PR, an inefficient dataloader is proposed.
These are the updates in this PR:

- new `PerClassOnlineDataset` in which samples are yielded only if  `label == self.filtered_label`. All the other samples are ignored. This attribute can be changed to filter on a different label
- support in the sample then batch function (grpc call to get all the distinct labels, _for_ loop to work label by label ..)
- support in the AbstractRemoteDownsampler (attribute `requires_data_label_by_label` and abstract function `_downsampler.inform_end_of_current_label` to handle the end of the label

### CODE REVIEW

File: modyn/selector/internal/selector_strategies/abstract_selection_strategy.py [hunk 1]
```diff
@@ -398,3 +398,20 @@ def _persist_samples(self, keys: list[int], timestamps: list[int], labels: list[
...
+                .filter(
+                    SelectorStateMetadata.pipeline_id == self._pipeline_id,
+                    SelectorStateMetadata.seen_in_trigger_id >= self._next_trigger_id - self.tail_triggers
+                    if self.tail_triggers is not None
+                    else True,
+                )
+                .distinct()
+                .all()
+            )
+            available_labels = [result_tuple[0] for result_tuple in result]
```
Comment (MaxiBoether) [inline modyn/selector/internal/selector_strategies/abstract_selection_strategy.py:415]: I suppose it is not a problem that these lists all labels that have been seen and it could be (e.g. due to random presampling) that not all labels are actually part of the TriggerTrainingSet in the end?

File: modyn/protos/selector.proto [hunk 2]
```diff
@@ -74,6 +76,14 @@ message GetNumberOfPartitionsRequest {
 
 message NumberOfPartitionsResponse { int32 num_partitions = 1; }
 
+message GetAvailableLabelsRequest {
```
Comment (MaxiBoether) [inline modyn/protos/selector.proto:79]: Could you please add an integration test for this :)?

File: modyn/selector/internal/selector_manager.py [hunk 3]
```diff
@@ -135,6 +135,12 @@ def get_number_of_partitions(self, pipeline_id: int, trigger_id: int) -> int:
 
         return self._selectors[pipeline_id].get_number_of_partitions(trigger_id)
 
+    def get_available_labels(self, pipeline_id: int) -> list[int]:
```
Comment (MaxiBoether) [inline modyn/selector/internal/selector_manager.py:138]: for this (and the other new functions), please add some unit tests as well. thank you!

File: modyn/trainer_server/internal/dataset/data_utils.py [hunk 4]
```diff
@@ -47,3 +48,19 @@ def prepare_dataloaders(
     # TODO(#50): what to do with the val set in the general case?
     val_dataloader = None
     return train_dataloader, val_dataloader
+
+
```
Comment (MaxiBoether) [inline modyn/trainer_server/internal/dataset/data_utils.py:52]: please mention issue 289 here :)

File: modyn/trainer_server/internal/dataset/per_class_online_dataset.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+            trigger_id,
+            dataset_id,
+            bytes_parser,
+            serialized_transforms,
+            storage_address,
+            selector_address,
+            training_id,
+        )
+
+        self._filtered_label = None
```
Comment (MaxiBoether) [inline modyn/trainer_server/internal/dataset/per_class_online_dataset.py:36]: any reason we do not set that in the constructor?

File: modyn/trainer_server/internal/dataset/per_class_online_dataset.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+            for idx, data_tuple in self._get_data_iterator(keys, data, labels, weights):
+                key, sample, label, weight = self._unpack_data_tuple(data_tuple)
+
+                if partition < self._num_partitions - 1 and idx == fetch_next_partition_idx:
+                    # TODO(#175) in case this blocks training
+                    new_keys, new_data, new_labels, new_weights = self._get_data(
+                        worker_id=worker_id, partition_id=partition + 1
+                    )
+                if label == self._filtered_label:
+                    yield self._yield_samples(key, sample, label, weight)
```
Comment (MaxiBoether) [inline modyn/trainer_server/internal/dataset/per_class_online_dataset.py:81]: hmm alternatively, couldnt we use the regular onlinedataset, and give it the optional parameter filtered label, and if it is not none, then we do this check? then we avoid introducing a new class. anything speaks against that?

File: modyn/trainer_server/internal/trainer/remote_downsamplers/abstract_remote_downsample_strategy.py [hunk 7]
```diff
@@ -62,3 +66,7 @@ def inform_samples(self, sample_ids: list[int], forward_output: torch.Tensor, ta
     @abstractmethod
     def select_points(self) -> tuple[list[int], torch.Tensor]:
         raise NotImplementedError
+
+    def inform_end_of_current_label(self) -> None:
```
Comment (MaxiBoether) [inline modyn/trainer_server/internal/trainer/remote_downsamplers/abstract_remote_downsample_strategy.py:70]: abstractmethod or not? you have a if that raises not implemented, not sure what this means.

File: modyn/trainer_server/internal/trainer/pytorch_trainer.py [hunk 8]
```diff
@@ -487,28 +495,32 @@ def downsample_trigger_training_set(self) -> None:
...
-                self._downsampler.inform_samples(sample_ids, model_output, target)
+        if self._downsampler.requires_data_label_by_label:
+            available_labels = self._get_available_labels_from_selector()
+            per_class_dataloader = prepare_per_class_dataloader_from_online_dataset(
+                self._train_dataloader.dataset, self._batch_size, self._num_dataloaders
+            )
+            number_of_samples = 0
+            batch_number = 0
+            for label in available_labels:
+                per_class_dataloader.dataset._filtered_label = label
```
Comment (MaxiBoether) [inline modyn/trainer_server/internal/trainer/pytorch_trainer.py:515]: ah now I see. usually a leading underscore indicates a private attribute. we should pass an initial filtered label in the constructor and make this attribute public

File: modyn/trainer_server/internal/trainer/pytorch_trainer.py [hunk 9]
```diff
@@ -533,10 +545,39 @@ def downsample_trigger_training_set(self) -> None:
...
 
         self.update_queue(AvailableQueues.DOWNSAMPLING, batch_number, number_of_samples, training_active=True)
+        # set the model to train
+        self._model.model.train()
+
+    def _iterate_dataloader_and_compute_scores(
+        self,
+        dataloader: torch.utils.data.DataLoader,
+        previous_batch_number: int = 0,
+        previous_number_of_samples: int = 0,
```
Comment (MaxiBoether) [inline modyn/trainer_server/internal/trainer/pytorch_trainer.py:555]: could you add some documentation why this is needed? it appears some per class vs non per class thing but its not obvious to me what is happening here

**Review by MaxiBoether — CHANGES_REQUESTED** [review]
Thank you! I had some small comments, I suppose after the next round we can merge :)

File: modyn/selector/internal/selector_strategies/abstract_selection_strategy.py (hunk 1 above)
Comment (francescodeaglio) [reply modyn/selector/internal/selector_strategies/abstract_selection_strategy.py:415]: Not a problem, simply no points are returned

Comment (francescodeaglio) [reply modyn/selector/internal/selector_strategies/abstract_selection_strategy.py:415]: To actually know which classes are selected, we could have to use more inefficient presampling methods (like `last used in trigger` as in the other PR). I don't think it is worth it.

File: modyn/trainer_server/internal/dataset/per_class_online_dataset.py (hunk 6 above)
Comment (francescodeaglio) [reply modyn/trainer_server/internal/dataset/per_class_online_dataset.py:81]: In my opinion, it is not worth changing the online dataset for a temporary feature. It is already complex enough, I would avoid adding to it.

I have now changed the iteration loop to 
```
data_tuple = self._get_data_tuple(key, sample, label, weight)
if data_tuple is not None:
   yield data_tuple
```

So we just have to overload the method `_get_data_tuple` and filter there

File: modyn/trainer_server/internal/trainer/pytorch_trainer.py (hunk 8 above)
Comment (francescodeaglio) [reply modyn/trainer_server/internal/trainer/pytorch_trainer.py:515]: Ok

File: modyn/trainer_server/internal/trainer/remote_downsamplers/abstract_remote_downsample_strategy.py (hunk 7 above)
Comment (francescodeaglio) [reply modyn/trainer_server/internal/trainer/remote_downsamplers/abstract_remote_downsample_strategy.py:70]: Weird inheritance. Now should be clearer

File: modyn/trainer_server/internal/dataset/per_class_online_dataset.py (hunk 5 above)
Comment (francescodeaglio) [reply modyn/trainer_server/internal/dataset/per_class_online_dataset.py:36]: Now toy can specify the initial label in the constructor and then change it at runtime

File: modyn/trainer_server/internal/dataset/per_class_online_dataset.py (hunk 6 above)
Comment (francescodeaglio) [reply modyn/trainer_server/internal/dataset/per_class_online_dataset.py:81]: So basically this is the only difference
```
def _get_data_tuple(self, key: int, sample: bytes, label: int, weight: Optional[float]) -> Optional[Tuple]:
    assert self.filtered_label is not None

    if self.filtered_label != label:
        return None
    return super()._get_data_tuple(key, sample, label, weight)
```

File: modyn/trainer_server/internal/trainer/pytorch_trainer.py (hunk 9 above)
Comment (francescodeaglio) [reply modyn/trainer_server/internal/trainer/pytorch_trainer.py:555]: Sure

File: modyn/protos/selector.proto (hunk 2 above)
Comment (francescodeaglio) [reply modyn/protos/selector.proto:79]: Added

File: modyn/selector/internal/selector_manager.py (hunk 3 above)
Comment (francescodeaglio) [reply modyn/selector/internal/selector_manager.py:138]: done

**Review by MaxiBoether — APPROVED** [review]
LGTM - 2 very minor comments, feel free to merge afterwards!

File: integrationtests/selector/integrationtest_selector.py [hunk 10]
```diff
@@ -624,10 +625,71 @@ def test_many_samples_unevenly_distributed():
     assert len(total_samples) == 10001
 
 
+def test_get_available_labels(reset_after_trigger: bool):
+    selector_channel = connect_to_selector_servicer()
+    selector = SelectorStub(selector_channel)
+    # We test without reset, i.e., after an empty trigger we get the same data
```
[the remark this answers is not in the document]
Comment (francescodeaglio) [reply integrationtests/selector/integrationtest_selector.py:631]: Yes, it was just `reset = false` at the beginning. Removed.

**Comment by MaxiBoether** [comment]
thanks for fixing, feel free to merge. note that robin just merged the evaluator, which might come in handy for evaluating the pipelines and collecting metrics :)
